src/Components/WebAssembly/WebAssembly/src/Rendering/WebAssemblyDispatcher.cs

```
@@ -0,0 +1,163 @@
+        _mainSyncContext = SynchronizationContext.Current!;
+
+        Debug.Assert(_mainSyncContext != null);
+        Debug.Assert(_mainSyncContext.GetType().FullName == "System.Runtime.InteropServices.JavaScript.JSSynchronizationContext");
+    }
+
+    public override bool CheckAccess() => SynchronizationContext.Current == _mainSyncContext;
+
+    public override Task InvokeAsync(Action workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (CheckAccess())
+        {
+            // this branch executes on correct thread and solved JavaScript objects thread affinity
+            // but it executes out of order, if there are some pending jobs in the _mainSyncContext already, same as RendererSynchronizationContextDispatcher
+            workItem();
+            // it can throw synchronously, same as RendererSynchronizationContextDispatcher
+            return Task.CompletedTask;
+        }
+
+        var tcs = new TaskCompletionSource();
+
+        // RendererSynchronizationContext doesn't need to deal with thread affinity and so it could execute jobs on calling thread as optimization.
+        // we could not do it for WASM/JavaScript, because we need to solve for thread affinity of JavaScript objects, so we always Post into the queue.
+        _mainSyncContext.Post(static (object? o) =>
+        {
+            var state = ((TaskCompletionSource tcs, Action workItem))o!;
+            try
+            {
+                state.workItem();
+                state.tcs.SetResult();
+            }
+            catch (Exception ex)
+            {
+                state.tcs.SetException(ex);
+            }
+        }, (tcs, workItem));
+
+        return tcs.Task;
+    }
+
+    public override Task<TResult> InvokeAsync<TResult>(Func<TResult> workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (CheckAccess())
+        {
+            return Task.FromResult(workItem());
```

this means if workItem throws, it will just bubble out instead of wrapping the exception in the Task return value. Is that intended? It feels like it would be better if exceptions always propagated out through the task

---

This is intentional for consistency with https://github.com/dotnet/aspnetcore/blob/0ed0bcdbfd0e0ca1617179e495ef893e36cb332c/src/Components/Components/src/Rendering/RendererSynchronizationContextDispatcher.cs#L46